src/librbd/api/Mirror.cc

```diff
@@ -415,18 +429,205 @@ struct C_ImageSnapshotCreate : public Context {
+  std::string image_header_oid = librbd::util::header_name(image_id);
+  cls::rbd::GroupSpec group_spec(group_id, group_ioctx.get_id());
+  cls::rbd::GroupImageStatus incomplete_st(image_id, image_ioctx.get_id(),
+				cls::rbd::GROUP_IMAGE_LINK_STATE_INCOMPLETE);
+  cls::rbd::GroupImageSpec spec(image_id, image_ioctx.get_id());
+
+  int r = cls_client::group_image_set(&group_ioctx, group_header_oid,
+				      incomplete_st);
+  if (r < 0) {
+    lderr(cct) << "couldn't put image into removing state: "
+	       << cpp_strerror(-r) << dendl;
+    return r;
+  }
+
+  r = cls_client::image_group_remove(&image_ioctx, image_header_oid,
+				     group_spec);
+  if ((r < 0) && (r != -ENOENT)) {
+    lderr(cct) << "couldn't remove group reference from image"
+	       << cpp_strerror(-r) << dendl;
+    return r;
+  } else if (r >= 0) {
+    ImageWatcher<>::notify_header_update(image_ioctx, image_header_oid);
+  }
+
+  r = cls_client::group_image_remove(&group_ioctx, group_header_oid, spec);
+  if (r < 0) {
+    lderr(cct) << "couldn't remove image from group"
+	       << cpp_strerror(-r) << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+void close_images(std::vector<I *> *image_ctxs) {
+  std::vector<C_SaferCond> on_finishes(image_ctxs->size());
+
+  for (size_t i = 0; i < image_ctxs->size(); i++) {
+    (*image_ctxs)[i]->state->close(&on_finishes[i]);
+  }
+
+  for (auto &on_finish : on_finishes) {
+    on_finish.wait();
+  }
+
+  image_ctxs->clear();
+}
+
+template <typename I>
+int open_group_images(IoCtx& group_ioctx, const std::string &group_id,
+                      std::vector<I *> *image_ctxs) {
+  std::vector<cls::rbd::GroupImageStatus> images;
+  int r = group_image_list_by_id(group_ioctx, group_id, &images);
+  if (r < 0) {
+    return r;
+  }
+
+  std::vector<C_SaferCond> on_finishes(images.size());
+  int ret_code = 0;
+  size_t i = 0;
+  for ( ; i < images.size(); i++) {
+    auto &image = images[i];
+    librbd::IoCtx image_io_ctx;
+    r = librbd::util::create_ioctx(group_ioctx, "image", image.spec.pool_id, {},
+                                   &image_io_ctx);
+    if (r < 0) {
+      ret_code = r;
+      break;
+    }
+
+    librbd::ImageCtx* image_ctx = new ImageCtx("", image.spec.image_id.c_str(),
+					       nullptr, image_io_ctx, false);
+
+    image_ctx->state->open(0, &on_finishes[i]);
+
+    image_ctxs->push_back(image_ctx);
+  }
+
+  for (size_t j = 0; j < i; j++) {
+    r = on_finishes[j].wait();
+    if (r < 0 && ret_code == 0) {
+      ret_code = r;
+    }
+  }
+
+  if (ret_code != 0) {
+    close_images(image_ctxs);
+    return ret_code;
+  }
+
+  return 0;
+}
+
+std::string calc_ind_mirror_snap_name(uint64_t pool_id,
+                                      const std::string &group_id,
+                                      const std::string &snap_id) {
+  std::stringstream ind_snap_name_stream;
+  ind_snap_name_stream << ".mirror." << std::hex << pool_id << "_"
```

Comment:
A better naming convention would be:
.mirror.<primary/non-primary>.<global_group_id>.<group_snap_id>